### DISCUSSION TITLE
Markdown Enhancements

### DISCUSSION THREAD

**Discussion by lorenzolewis** [withastro/roadmap#424] [Proposal]
# Background

I absolutely love the functionality and direction that the [Content Collections RFC](https://github.com/withastro/rfcs/blob/main/proposals/0027-content-collections.md) guides Astro towards. It's a lot of really good baseline work that will hopefully help those integrating content that might not be written alongside an Astro project.

That being said, there can be a lot of friction to a develop new to Astro coming from another platform where they're expecting enhanced markdown functionality out of the box. These can often times be fixed by changing the format of the Markdown, but those are often seen as workarounds and may not be possible to do (such as when a developer's Markdown content is being rendered to another service in parallel and they don't have the flexility to change it).

# Current Behaviour

There are already some defaults applied to all markdown parsed by Astro: https://docs.astro.build/en/guides/markdown-content/#configuring-markdown-and-mdx

# Missing Behaviour

The below list contains functionality that I'm personally missing. I'm sure there are a lot more and that some are more relevant than others.

- Assets relative to Markdown files (https://github.com/withastro/rfcs/discussions/334, https://github.com/withastro/rfcs/pull/381, and https://github.com/withastro/rfcs/discussions/23 all touch on this as well as the various issues referenced within them)
- Link handling and resolution between Markdown files (https://github.com/withastro/astro/issues/5682 and https://github.com/withastro/astro/issues/5680)
- Mermaid diagram support (https://github.com/withastro/rfcs/discussions/267)
- Internal broken link detection (and ability to error during build) similar to https://docusaurus.io/docs/api/docusaurus-config#onBrokenMarkdownLinks
- `readme.md` being used as an index page
- Extract classes from image references to support use cases such as [GitHub-style themed images](https://github.blog/changelog/2021-11-24-specify-theme-context-for-images-in-markdown/)
- Note admonitions for quote blocks in the style of https://github.com/community/community/discussions/16925

Not all of these may make sense and by introducing more it opens up a can of worms for a potential flood of "well just add xyz feature" for the way Astro parses Markdown.

# Current Solutions

Right now you can add in your own remark and rehype plugins as outlined in https://docs.astro.build/en/guides/markdown-content/#markdown-plugins. This can be a daunting task for those who might not be familiar with these tools (how many times did I have to look up the difference between what remark and rehype do today because I kept forgetting? 😅).

There is a very large plugin ecosystem for remark and rehype, but again it can get daunting knowing which plugin to use, how it affects the rendered content, and exposes a whole new level of complexity to a developer. This can cause a lot of DX friction for those moving from another platform where they're used to some of these things happening by default (moving from Docusaurus in my case).

# Proposed Solutions

1. Add opt-in behaviour for "enhanced" markdown parsing/rendering as a universal Astro config
2. Add opt-in behaviour per Content Collection to selectively turn this on per collection
3. Curate a list of community- and/or Astro-maintained remark/rehype plugins that map to common use cases someone might expected when coming from another platform

I'd personally prefer either option 1 or 2. It would reduce the friction of developers coming from another platform where they're used to these niceties. This can be a set of "blessed" functionality that I can count on working in Astro and not have to worry about community plugins, updating additional dependencies, wondering how they all interact with each other, etc.

---

Hopefully there's more feedback on this as I'd love to get the perspective of those in the community who have been around longer than I have; I'd love to hear their feedback! 

Oh, and thanks for the amazing Astro project <3

**Comment by lorenzolewis** [comment]
One option that comes to mind around what the configuration would look like is inspired by https://github.com/withastro/astro/pull/5684.

There could an enum in the config that contains these values:

- `none`: Astro does no processing with the markdown files
- `basic`: (default) Astro does processing like it does today (with GitHub Flavoured Markdown, code highlight, heading ID generation)
- `enhanced`: This could cover the cases mentioned in the original discussion post (at least the ones chosen to include). 

This would also bake in the functionality where a user really doesn't want Astro to do any processing on their files.

**Reply by delucis** [reply]
Interesting ideas here! And quite a few of your “Missing Behaviour” list are things we needed to build custom solutions for for the Astro Docs site, so would definitely be interesting to think about generalizing those.

Pretty much all of these exist in the remark/rehype plugin space, so I wonder if this can be an integration — even potentially a userland experiment initially — rather than built in.

How exactly a user opts in/out of things is also something to resolve. While the `none`/`basic`/`enhanced` is simple to get started, it falls apart once a user says “Love `enhanced` but how do I disable feature X?” (This is kind of the issue we have today with Astro’s defaults.) But that can be tackled later.

**Reply by lorenzolewis** [reply]
It could be that `none | basic | enhanced` are communicated as a collection of plugins that help someone get the general functionality they want. But then if we wanted to give more granular abilities to tweak them then you use the `remarkPlugins` config for that (to either replace or extend the current plugins. 

I'm probably going to build these out in the tauri.app rewrite we're doing in Astro and plan to keep each plugin separate. If we want to continue down this then could look at pulling them upstream directly into Astro.

Like you said, a lot of these already exist in the plugin space, but I think there could be a benefit to bundling/redistributing them with Astro for the sake of a user not having to dig into that ecosystem if they choose not to.

**Reply by bholmesdev** [reply]
I agree with your point there @delucis. Admittedly, I'm not comfortable using blanket terms for "basic" and "enhanced," as it's fairly opaque what is applied under each setting. I also don't think a simplified enum allows enough [inversion of control](https://kentcdodds.com/blog/inversion-of-control) to pick-and-choose presets you need. I think the target demographic for "enhanced" users aren't those that just want _more_ features; they usually need a preset of options suited to their use case, like "enhanced for blogs" or "enhanced for docs."

To solve this, I could imagine each of these being "preset" objects rather than string values. You can either apply outright or choose properties individually:

```ts
import { basic, enhanced } from '@astrojs/markdown-remark/presets';

markdown: {
  // 1. Apply preset untouched
  preset: enhanced,
  // 2. Choose pieces you need
  preset: {
    blockquotes: enhanced.blockquotes,
  }
}
```
This is definitely doable from an Astro integration today though! I'd love to see user feedback on that.

**Reply by lorenzolewis** [reply]
I like the code example @bholmesdev. 

Another options is that maybe `basic` and `enhanced` are just a re-exported array of plugins. Then you could use them like this:

```ts
import { enhancedPreset, blockquotes } from '@astrojs/markdown-remark';

markdown: {
  // a. Use a preset by spreading it
  plugins: {[
    ...enhancedPreset
  ]},
  // b. or cherry pick the specific ones you would like
  plugins: {[
    blockquotes
  ]}
}
```

This leaves out the distinction between remark vs rehype plugins of course, but gets the general idea across.

**Reply by delucis** [reply]
Yeah, I think the tricky bit is actually handling what’s remark and what’s rehype if you’re relying on folks using this directly inside the different `re____Plugins` arrays. Making this an integration can help solve for that if you’re comfortable with hiding the logic a bit: use an object like in Ben’s sketch and the integration decides behind the scenes what ends up in which plugins array.

**Comment by jonsage** [comment]
Just adding a couple of cents.

With regard to link resolution of .md files, my preferred behaviour would be that if I have relative (or absolute) links to files in the static source, then those links would magically be transformed to continue working in the output HTML.

I would think Astro knows the original file paths of all files, so it could be some functionality, probably hooked into reMark plugins but with config from Astro, that when processing links, checks them against the graph and grabs the permalink of each and swaps that in.

It's probably not straightforward, but it means markdown authored to work on github or in vscode or other tools could be built to html and continue working - otherwise it means links will be broken in one side or the other.

Jekyll has a plugin for such functionality with an explanation behind it's reasoning:

https://github.com/benbalter/jekyll-relative-links

I would like to see this as an option. The purist in me desires the simplicity of a folder of inter-linked markdown files that can be browsed as source (excellent for docs sites where people can browse in GitHub) as well has transformed into a beautiful ~~butterfly~~ website.

**Comment by vernak2539** [comment]
In regards to the "resolving relative assets" part, I've built a rehype plugin to satisfy my want to link between markdown/mdx files to other markdown/mdx files using relative paths. It can be found on [Github](https://github.com/vernak2539/astro-rehype-relative-markdown-links) and [NPM](https://www.npmjs.com/package/astro-rehype-relative-markdown-links).

Hopefully it can serve as a stopgap until the functionality is decided to be put into Astro (or can be used as a user-land thing if not).

**Reply by bholmesdev** [reply]
Thanks for sharing this! Today, the experimental assets flag should resolve relative images through our image optimizer, using relative or alias paths. Curious if this solves your issue as well: https://docs.astro.build/en/guides/assets/

**Reply by vernak2539** [reply]
I didn't even consider trying that out. I will give it a shot hopefully in the next couple days and see how it goes. Thanks!

**Reply by vernak2539** [reply]
Hey @bholmesdev, I was able to try this out yesterday. I enabled the experimental assets feature and disabled my plugin. Unfortunately, I did not work, and I was left with links to other markdown/mdx files when using `astro dev` and `astro build`. 

Based on the docs, it does look like this functionality could be expanded to markdown/mdx links as well.

Is there any chance that you could point me to the code where this is implemented (sorry, being a bit lazy during my day job ha)? Maybe I could have a look at what this could look like inside of Astro 🤷

**Reply by delucis** [reply]
Ah, I think there were some crossed wires here — the new assets solution is for _images_ not for links between pages/Markdown files.

**Reply by vernak2539** [reply]
Ahh okay! No worries, thanks for clarifying! Would've been cool though amirite?!?! ha

**Reply by bholmesdev** [reply]
Ah, misunderstood the ask then! I’m a bit confused what you _would_ expect to happen for Markdown “links.” Do you mean anchor links to relative Markdown paths, where those relative links become live URLs on your site? If so, how would you expect layouts to work?

**Reply by vernak2539** [reply]
No worries. I guess the ask relates to the bullet point above (I realise I wasn't clear):

>Link handling and resolution between Markdown files (https://github.com/withastro/astro/issues/5682 and https://github.com/withastro/astro/issues/5680)

For example, if you have a markdown file in a content collection at `src/content/blog/post.md` that defines a link like below:

```md
[relative link](./other-markdown.md)
```

By default, I would (likely) want resulting HTML to be:

```html
<a href="/blog/other-markdown">relative link</a>
```

This is what the rehype plugin I made does (transforms the link in the AST -> HTML phase) and it's something that looks very similar to the experimental assets functionality you linked prior

**Reply by bholmesdev** [reply]
Ah, this would indeed be more opinionated. When using content collections specifically, your content is detached from the URL that renders your content. So a plugin wouldn't know what route to generate! We'd need a clearer story on direct content -> URL mappings to make that jump.

**Reply by vernak2539** [reply]
>When using content collections specifically, your content is detached from the URL that renders your content.

That's super interesting! Would you mind pointing me to more information about this?

I feel like I'm missing something (again still new to Astro, content collections, etc.), but I thought the final path to generate would based on the relative path to content would depend on:
1. The filesystem / directory based routing (isn't this what Astro uses?)
2. The slug in the frontmatter

And as I wrote this, I may have realised where I'm lacking, which would be great if you could confirm. The problem I'm likely "glancing" over is the fact that in my example `src/pages/blog` would have to be set up correctly. It would need to likely contain a `[...slug].astro` file and/or other files that handle the routing.

Using [my blog](https://github.com/vernak2539/words-byvernacchia) as an example, the plugin I created works because of the pages I have defined in the `src/pages/blog`?

Or... maybe I've missed the mark completely. Looking forward to discussing.

**Reply by shishkin** [reply]
I think the issue is that content collections [rely on a dynamic route](https://docs.astro.build/en/guides/content-collections/#generating-routes-from-content) to be rendered. So the dynamic page needs to use URL params to lookup the item in the content collection and render it. This function can be arbitrary and thus is not generally reversible. And the relative link resolver needs to go the other way around from a local relative file to URL params. But it should definitely work for [static routing](https://docs.astro.build/en/core-concepts/routing/#static-routes) of MD pages.

**Reply by bholmesdev** [reply]
@vernak2539 Exactly right! You'll use dynamic routes to map content collections, or you can store in `src/pages/` directory. The latter approach lets you do magical route generation for relative `.md` paths if this is what you prefer. Though in practice, we prefer explicit dynamic routes to make route generation clear. This is admittedly why we've ruled out relative `.md` path resolution for the time being 🤷

**Reply by vernak2539** [reply]
That makes a lot more sense now. I've updated my package with all of the assumptions I made prior to this thread. Will very much look forward to anything else that comes out about relative linking inside markdown content (aside from assets). Thanks again!

**Reply by bholmesdev** [reply]
Love these discussions! I'll take a look at the plugin to see what Astro could learn from ❤️

**Reply by stereobooster** [reply]
2¢ to this discussion. [Portable markdown links](https://stereobooster.com/posts/portable-markdown-links/) are supported by: Github, VSCode, Hugo, Docusaurus, Obsidian. It would be nice if Astro can support it.
